src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -462,30 +460,32 @@ internal void Flush()
             /// </summary>
             private void RaiseFunctionMetricEvents()
             {
-                if (_runningFunctions.Count == 0)
+                if (_runningFunctions.IsEmpty)
                 {
                     return;
                 }
 
                 // We only need to raise events here for functions that aren't completed.
                 // Events are raised immediately for completed functions elsewhere.
-                FunctionStartedEvent[] runningFunctionsSnapshot = null;
-                lock (_runningFunctionsSyncLock)
+                // Loop through and prune any completed runs
+                foreach (var possiblyRunning in _runningFunctions)
                 {
-                    // effectively we're pruning all the completed invocations here
-                    _runningFunctions = _runningFunctions.Where(p => !p.Completed).ToList();
-
-                    // create a snapshot within the lock so we can enumerate below
-                    runningFunctionsSnapshot = _runningFunctions.ToArray();
+                    if (possiblyRunning.Value.Completed)
+                    {
+                        _runningFunctions.TryRemove(possiblyRunning.Key, out _);
+                    }
                 }
 
-                // we calculate concurrency here based on count, since these events are raised
+                // If not for the length, we could probably avoid this array allocation
+                var concurrency = _runningFunctions.Count;
+
+                // We calculate concurrency here based on count, since these events are raised
                 // on a background thread, so we want the actual count for this interval, not
                 // the current count.
                 var currentTime = DateTime.UtcNow;
-                foreach (var runningFunction in runningFunctionsSnapshot)
+                foreach (var runningFunction in _runningFunctions.Values)
```

Comment:
Values also acquires the locks and creates a new collection/snapshot. Is that necessary, or could another enumeration be good enough?

Reply:
We could get more complicated here, but we'd also quickly get into diminishing returns vs. maintenance of something more complex, IMO. While `GetValues()` under the hood does a lock acquisition and snapshot, we're generally talking about very few items here. The cost amortization is also important - this runs once every `n` seconds to dump based on config but with the new cleanup only contains the running values (+/- instantaneous delta inside any method), then we're only locking _once per cycle_ rather than a lock on every op (where we see a great deal of contention in profiling).

Talking through this though - we _could_ change it up. We don't need _another_ enumeration though, we've already got one above. If we're creating the same list of items _anyway_ we could just pretend we're microseconds earlier and go with an `else` in the loop above, creating our "running" collection in the existing enumeration. It'd eat the same number of resizes in the list (still relatively rarely in the overall scheme) as the current `.Values` call.

Reply:
@danielmarbach updated :)